### DISCUSSION TITLE
Specification pattern discussion

### DISCUSSION THREAD

**Discussion by josephdangerstewart** [elsa-workflows/elsa-core#1224] [Design & tech]
Hello again! 👋 

I have a more immediate issue regarding non-Linq based definition stores that I need a solution for and a more general architecture discussion about the specification pattern. To jump to the practical discussion, just skip the first two headings.

I'd like to try and make the case that the specification pattern (introduced [here](https://github.com/elsa-workflows/elsa-core/pull/517)) actually makes implementing a persistence provider harder and politely prompt consideration of switching back to the 1.x persistence interface shape.

My argument is two fold:

### 1. The specification pattern leaks implementation details

The fact that the default persistence providers are databases with a Linq-to-database-query API should be (in my opinion) an implementation detail of those providers. The fact that the persistence interfaces take in Linq expression `WHERE` clauses leaks that implementation detail. This precludes something like using an external REST API for a workflow definition store, unless that API also exposes a hyper-generic surface. But even then, I would have to implement a Linq-to-api-request mapper which is not a trivial task, and it would be difficult to optimize my API requests for Elsa's usage.

Furthermore, it makes it difficult for me to store workflow definitions in a different shape that can be mapped Elsa's model. I can load all workflow definitions into memory, map them to Elsa's model, and *then* run a `.Where(specification.ToExpression())` on the results, but that won't scale to our use case.

### 2. The specification pattern doesn't provide a contract for implementers to optimize for

As a corollary of the above, the interface contract doesn't tell me anything how Elsa is actually using these methods. I don't know from the interface what kinds of queries I should be optimizing my implementation for. The [previous interface](https://github.com/elsa-workflows/elsa-core/blob/1.4.x/src/core/Elsa.Abstractions/Persistence/IWorkflowDefinitionStore.cs) gives me a comprehensive list of queries I need to optimize for as a persistence implementer.

Essentially, my point is that the scope of the current interfaces is not rigid enough to be reasonably implemented for non-Linq based storage providers.

### My use case for context

The reason I bring this up at all is because the company I'm working for is building a workflow feature in our platform and we're using Elsa as a workflow execution engine (see #1190). We've decided to implement our own API for CRUD operations on our own workflow definition model that can be mapped to Elsa's model. In my Ideal World™️, I would be able to write a workflow definition service that is the single source of truth for workflow definitions. Our controllers would use this service for all CRUD operations on workflow definitions. I could then implement my own `IWorkflowDefinitionStore` that uses that same workflow definition service to provide workflows to Elsa. I would like to avoid allowing consumers of my workflow service to provide arbitrary `WHERE` clauses via Linq expressions mostly for performance reasons (I'm sure there are security reasons as well that I would have more readily available if I were more familiar with Linq expressions and EF). For instance, I would like my service to not allow listing all workflows (a problem I had a workaround for in 1.x).

### Alternative questions to address my immediate problem

My plan right now is to just use Elsa 1.x for our project although I'd love to find a workaround for 2.0 to get more of the benefits of an actively developed open source project.

1. Do I even need to be implementing my own persistence provider? Can I trigger workflows and have Elsa execute on them without ever needing to query the definition store?
    1. My guess is no since [`WorkflowInstances`](https://github.com/elsa-workflows/elsa-core/blob/master/src/core/Elsa.Abstractions/Models/WorkflowInstance.cs#L19) store a reference to a definition id
2. If Elsa does need a workflow definition store, is there a service higher up in Elsa's architecture that I can override instead of `IWorkflowDefinitionStore` that has a more tightly defined scope?
3. Is there an easy hack that you are aware of for implementing a non-Linq based definition store that I'm missing?

Tangentially related: Loading all workflow definitions in memory is not going to scale for us. In 1.x, my plan was to add my own implementation of [`IWorkflowRegistry`](https://github.com/elsa-workflows/elsa-core/blob/16611e2a744933aa478d3fae9e995917e2ba5dbf/src/core/Elsa.Abstractions/Services/IWorkflowRegistry.cs) that didn't load workflow definitions into memory. As far as I could tell, that was the only service in the 1.x core service code that was actually calling `ListAsync` on the `IWorkflowDefinitionStore` (my plan was to just throw a `NotImplementedException` from that method in my implementation). I haven't dug as much through the 2.0 code but do you know what it would take to prevent loading all workflow definitions into memory in 2.0?

**Comment by sfmskywalker** [comment]
Hi @josephdangerstewart,

Let's go over each heading.

### 1. The specification pattern leaks implementation details

Indeed it does; not out of necessity, but out of being helpful without being a requirement. 
The reasoning here is that a specification describes a query. It is up to the implementer to interpret this specification as they see fit.

Most providers have a LINQ provider in place, so having specifications that have a default expression simplifies data access queries for most providers. Is it 100% pure? Nope. But it is practical without posing any limitations on the provider implementing it.

The provider simply has to check which specification is being handled, and act in any way that makes sense for the specification.

This does not preclude providers that use something like an external REST API for a workflow definition store. Such providers might do something like this:

```csharp
public async Task<T?> FindAsync(ISpecification<WorkflowDefinition> specification, CancellationToken cancellationToken = default)
{
   if(specification is EntityIdSpecification<WorkflowDefinition> entityIdSpecification)
      return await _httpClient.GetWorkflowDefinitionByIdAsync(entityIdSpecification.Id, cancellationToken);
  // Handle all the specifications.
}
```

The YesSQL provider takes advantage of this by only using the expressions that can directly be mapped to its internal document model, but in those cases where the models aren't compatible, it handles the specification explicitly.

This also allows you to use any shape of models you like. All the `IStore` abstraction needs you to do is:

1. Accept some specification, and 
2. Return the requested object.

It doesn't care how you persist or from where you retrieve the objects. The `ToExpression` is there for convenience and therefore optional.

An obvious downside of course is that if a developer decides they want to add another specification, they will have to subclass all providers that can't use the LINQ expression and handle the specification implicitly.

The upshot of the specification pattern is that in many cases, one can add a specification and it will work without touching the providers.

### 2. The specification pattern doesn't provide a contract for implementers to optimize for

I don't think that's true. As outlined above, your implementation is perfectly capable of optimizing a query. The most important difference between a "rigid" contract versus a "specification" contract is that the specification contract is open-ended in terms of what queries it can support. But it does not say anything about how you want to implement these specifications. 

The same goes for a rigid contract - although its number of operations are finite, it's up to the implementation to do the right thing.

That being said, there are definitely problems with the (current) specification pattern implementation, but for other reasons.

Specifically, it makes you think specifications can be combined into e.g. AND and OR operations. But that only works when relying on the (optional) expressions associated with a specification. If you wanted to support this in an actual implementation that is unable to automatically translate the final expression into e.g. API calls, then you' be out of luck and you would have to try and handle all of the permutations of combinations of specifications.

For example, if we updated some service to query something by combining an existing specification with another one and joining them using AND, your API persistence provider would now be insufficiently implemented and break the system, since it wouldn't even be aware of this new combined specification.

Obviously, if we were using rigid contracts, I would break your implementation too when I added a method. But at least you would know about this at compile time when the build fails after updating your packages to their latest versions :)

### My use case for context

You mentioned that you want to prevent your consumers from using arbitrary LINQ expressions. Are these consumers developers that integrate with your own .NET library? If so, wouldn't they be able to "hack" the system anyway, by e.g. accessing the underlying persistence store directly?

### Alternative questions to address my immediate problem

1. If you have your own version of a Custom Workflow Definition Service and your own Workflow Definition Model, then implementing your own persistence provider that uses this Custom Workflow Definition Service makes sense. Of course I know too little about your exact requirements, but another approach might be to have your custom workflow model **reference** the Elsa workflow model by ID. Basically, your custom model would store anything that your application needs, plus a workflow definition ID that points to a workflow stored by Elsa. When you display the workflow designer, that the user opened by clicking some "Edit Workflow" link in your portal, your application will first load the custom workflow model and then continue on to the workflow editor page with the Elsa Workflow Definition ID provided as a parameter. This way, you could even treat Elsa Server as a separate component in your architecture (but you don't have to).

2. You only need a definition store if you want to be able to manage workflow definitions. The engine, however, doesn't care about workflow definitions - it only cares about **Workflow Blueprints**, which are derived from various sources - the workflow definition store is just one of them. Another source for example are workflows built in C# using the Workflow Builder API. Another provider is a JSON based provider that reads JSON files from blob storage using Storage.NET - an abstraction of various file systems, including hard drives, Azure Blob storage and FTP to name just a few. You could implement your own provider that reads from your own API that provides custom workflow models - all it needs to do is map these to actual `WorkflowBlueprint` objects, and you're set.

3. As you might have guessed by now: absolutely: the easiest hack is to handle the various specifications explicitly using e.g. some switch expression that maps specification objects to API calls. A more maintainable approach might be to have a set of specification handlers, where you simply lookup a handler that can handle a given specification type.

Regarding Workflow Registry having all workflows in-memory, I'm with you on that - although all workflow blueprints are currently held in memory, it should be doable to reconstruct them on demand (keeping in mind that not all workflow blueprints are derived from workflow definitions, but also from C# classes etc.) thanks to the fact that we have workflow triggers - tiny objects that store details about what workflow blueprint to run and what activity to start should a given condition occur. These workflow triggers too are stored in memory, but there's no reason they couldn't be stored in a DB or other storage mechanism.

If you want to pursue this further, I recommend you create separate issues so we can discuss any further details and plan for them etc.

As a final note, I'm open to revisiting the provider implementation to a more rigid set of contracts. If you feel strongly about this, then I would love to see a proposal that takes into account the current set of operations necessary (of which are more than there are in Elsa 1).

**Reply by josephdangerstewart** [reply]
@sfmskywalker, thank you for your very well thought out and detailed response as well as your suggestions on my more immediate problem! I'll start by responding to your suggestions regarding my more immediate problem because the rest of this response is more theoretical than practical.

> The engine, however, doesn't care about workflow definitions - it only cares about Workflow Blueprints

Wonderful, this was just what I need to be pointed in the right direction. Do you think it would make sense to implement my own workflow registry and throw `NotImplementedException`s from all of the methods except `GetAsync`?

For more context, we're currently planning on using Elsa purely as a workflow execution engine. For starting workflows, I am now planning on querying our API directly and calling [`StartWorkflowAsync`](https://github.com/elsa-workflows/elsa-core/blob/725bb5e7c74ac3b64440ae7789c0828b3904b34f/src/core/Elsa.Abstractions/Services/Workflows/IStartsWorkflow.cs#L9) directly on the mapped result. For resuming workflows, I am now planning on calling [`FindAndResumeWorkflowsAsync`](https://github.com/elsa-workflows/elsa-core/blob/725bb5e7c74ac3b64440ae7789c0828b3904b34f/src/core/Elsa.Abstractions/Services/Workflows/IFindsAndResumesWorkflows.cs#L9) using `correlationId` to scope instances to just workflows in a particular group (so that not all workflow instances are loaded into memory). After looking through the code paths for those two methods, it only looks like `GetAsync` is called from the workflow registry service.

Now onto the discussion about the specification pattern. Unfortunately, I don't have the time to put together a more formal alternative proposal that takes into account the current usage of the definition/instance stores, but I did have some thoughts on your responses. Although since my immediate problem is now tentatively solved, this is more for the sake of interesting discussion so feel no need to continue it if you don't find it fruitful.

### 1. The specification pattern leaks implementation details

> The provider simply has to check which specification is being handled, and act in any way that makes sense for the specification.

I think this statement actually gets to the heart of what I mean and why I think it could pose a problem currently for customizability and in the future for maintainability. The interface contract would suggest that I can pass in any object that implements [`ISpecification<T>`](https://github.com/elsa-workflows/elsa-core/blob/725bb5e7c74ac3b64440ae7789c0828b3904b34f/src/core/Elsa.Abstractions/Persistence/Specifications/ISpecification.cs) and get a sensible result. This works fine for EF (and EF-like) storage mechanisms because I can pass the return value of `ToExpression()` directly as my query. This also works well for smaller storage mechanisms (such as memory or disk) where I can load all entities into memory and call `results.Where(specification.IsSatisfiedBy)`.

All other implementations would have to rely on checking for particular implementations of `ISpecification<T>` (which, if I'm understanding you correctly, is what you're suggesting and doing in the YesSQL store). However, doing so would be going against what I, as a consumer of `IStore<T>`,  would expect. As a consumer of `IStore<T>` I would expect to be able to pass in my own implementation of an identity specification and see no difference in behavior.

Of course, this may not matter right now because I'm *not* going to pass in my own implementations of `ISpecification<T>` and I could probably find a compressive list of specification types that could be passed into `FindAsync` et al. *But*, this does make me nervous for continued development of the project because what could be a minor feature, patch, or refactor ends up being a breaking change without *technically* changing the API contract.

> This does not preclude providers that use something like an external REST API for a workflow definition store

The current implementation of the specification pattern doesn't preclude this, but the interface contract still does because I can't pass in any implementation of `ISpecification<T>` into `FindAysnc` and get a list of all matching results. I can only handle particular implementations of `ISpecification<T>`.

> But at least you would know about this at compile time when the build fails after updating your packages to their latest versions :)

This is true! My custom implementation breaks when you add a new method either way. Although I would prefer this kind of a change
A) be very hard to make unintentionally and
B) be very very loud if it does need to be made (preferably with an accompanying major version bump)

And to extrapolate a more general principle, I think this is why, in general, it's dangerous for an abstraction to expose implementation details - because things don't fail loudly when those implementation details change.

### 2. The specification pattern doesn't provide a contract for implementers to optimize for

> As outlined above, your implementation is perfectly capable of optimizing a query.

I agree with you that this is still possible, although the interface contract doesn't give me any direction about how to do this or that my optimizations will still work (even if the code technically does) when certain implementation details change (e.g. a new implementation of `ISpecification<T>` is added). My point here was not about it being impossible to optimize so much as it was about it being unreliable given the `IStore<T>` interface contract.

> Specifically, it makes you think specifications can be combined into e.g. AND and OR operations

This is also part of what I was trying to get at, the current interface doesn't tell me if this I should be expecting this or not. Because this is possible given the interface contract, however, I wouldn't have a complete implementation if I *didn't* handle it (which is undesirable given my storage method). But to optimize for the general case, I have to look at how these services are being used, and even then, there's no guarantees in the interface contract that this wont change in the future (indeed, it seems to be the point of the pattern).

So in conclusion, even if I can map an arbitrary LINQ expression to an API call, any optimizations I make can only work for how these interfaces are currently being used, not how you expect to be able to use them in the future.

For the time being, however, I'm happy to rely on the hack I outlined above especially given the reduced scope in which we're actually using Elsa.

**Reply by sfmskywalker** [reply]
Hi @josephdangerstewart ,

>Do you think it would make sense to implement my own workflow registry and throw NotImplementedExceptions from all of the methods except GetAsync?

Even better: all you have to do is implement `IWorkflowProvider` which has a single method to implement: `GetWorkflowsAsync`. Have it return workflow blueprints, register your workflow provider with DI, and you should be set.

Thank you for your additional thoughts and feedback on the specification pattern. I believe we're on the same page in terms of the disadvantages and I am leaning towards revisiting this for Elsa 3. Which gives us some time to further reflect upon a better story for data access.

## (Un)related Topic
On a slightly related note yet unrelated to the specification pattern: the biggest challenge I'm finding is to find a way for Elsa to have a pluggable storage mechanism like we have today, but also allow us to build "modules" that plugin to Elsa. Some of these modules might introduce new entities, which require their own persistence abstractions in order for them to be pluggable. 

A good example of this is the Webhooks module, which introduces the concept of a Webhook Definition.

If Elsa used a single ORM like Entity Framework Core, then it would be easy for such modules to simply define their own db context and leverage EF Core's APIs.

But instead, it now it has to be decided whether the Webhooks module should implement the same set of persistence providers as Elsa Core, such as EF Core, MongoDB, YesSQL and more to come.

I don't actually expect to find a solution to this, as it seems to me it boils down to making a choice between keeping Elsa's persistence layer abstract so that it remains pluggable, or remove the abstraction and expose some ORM throughout the code base so that Elsa can be easily extended with modules that also require DB access (without having to duplicate a lot of persistence providers).

**Reply by josephdangerstewart** [reply]
> Even better: all you have to do is implement `IWorkflowProvider` which has a single method to implement:` GetWorkflowsAsync`. Have it return workflow blueprints, register your workflow provider with DI, and you should be set.

Unfortunately, any solution that requires loading all workflow definitions into memory is prohibitive for us. But I was able to get things set up with my own implementation of `IWorkflowRegistry`, I realized as I was testing it that `ListActiveAsync` is called [here](https://github.com/elsa-workflows/elsa-core/blob/master/src/core/Elsa.Core/Services/Bookmarks/BookmarkIndexer.cs#L64) in `BookmarkIndexer`. So for now, I just have `ListActiveAsync` returning an empty list. My assumption is that this means bookmarks wont be saved for my workflow instances and I wont be able to find workflow instances by bookmark. Is that correct?

> Thank you for your additional thoughts and feedback on the specification pattern.

Of course, thank you for your time and for your help.

> without having to duplicate a lot of persistence providers

I would ask why duplicate persistence providers is a negative? It seems to me that all of these different entities are pretty different from each other with pretty different requirements. And if these entities are different from each other, a shared storage abstraction seems like the wrong direction.

There are definitely aspects of these entities that are the same by design. Those aspects could be abstracted away in a common `IStore<T>` interface, but (#in-my-opinion) the methods that are unique to each entity storage should be unique to the persistence interface. Here's an example of what I mean

```cs
public interface IStore<T> {
    Task<T> GetByIdAsync(string id);
    Task<T> CreateAsync(T entity);
}

public interface IWorkflowInstanceStore : IStore<WorkflowInstance> {
    // Listing by correlation id is a requirement unique to workflow instances
    Task<IEnumerable<WorkflowInstance>> ListByCorrelationId(string correlationId);
}
```

Of course, this approach may seem less flexible since you have to add a new method to this interface anytime you need to retrieve workflow instances in a new way. Although I would argue that the current pattern is only more flexible when you're using entity framework (or some Linq based storage mechanism). As we've already discussed, for other storage mechanisms adding a new specification is just as much of a breaking change as adding a new method to the `IWorkflowInstanceStore` interface. The only difference is that adding a new method to the interface will explode loudly when I try to build instead of causing subtle bugs or runtime errors in production.
